src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,365 @@
+enum
+{
+	ClientPhase_WAITINGSERVERDATA = 2,
+	ClientPhase_WAITINGSERVERCOMPLETE = 3,
+	ClientPhase_PENDINGCOMPLETE = 4,
+	ClientPhase_COMPLETE = 5,
+};
+
+/** Server handshake state phases. */
+enum
+{
+	ServerPhase_WAITINGCACK = 2,
+	ServerPhase_COMPLETE = 3,
+};
+
+
+
+
+
+cForgeHandshake::cForgeHandshake(cClientHandle *client) :
+	m_IsForgeClient(false),
+	m_Errored(false),
+	m_Client(client)
+{
+}
+
+
+
+
+
+void cForgeHandshake::AugmentServerListPing(Json::Value & a_ResponseValue)
+{
+	UInt32 ProtocolVersion = m_Client->GetProtocolVersion();
+	cForgeMods & Mods = cRoot::Get()->GetServer()->GetRegisteredForgeMods(ProtocolVersion);
+
+	if (!Mods.HasMods())
+	{
+		return;
+	}
+
+	LOG("Received server ping from version: %d", ProtocolVersion);
+
+	// modinfo:
+	Json::Value Modinfo;
+	Modinfo["type"] = "FML";
+
+	Json::Value ModList(Json::arrayValue);
+
+	for (auto & item: Mods.GetMods())
+	{
+		Json::Value Mod;
+		Mod["modid"] = item.first;
+		Mod["version"] = item.second;
+		ModList.append(Mod);
+	}
+
+	Modinfo["modList"] = ModList;
+
+	// Augment the response:
+	a_ResponseValue["modinfo"] = Modinfo;
+}
+
+
+
+
+
+void cForgeHandshake::BeginForgeHandshake(const AString & a_Name, const AString & a_UUID, const Json::Value & a_Properties)
+{
+	ASSERT(m_IsForgeClient);
+
+	m_Name = &a_Name;
+	m_UUID = &a_UUID;
+	m_Properties = &a_Properties;
+
+	std::array<AString, 5> Channels{{ "FML|HS", "FML", "FML|MP", "FML", "FORGE" }};
+	AString ChannelsString;
+
+	for (auto & Channel: Channels)
+	{
+		ChannelsString.append(Channel);
+		ChannelsString.push_back('\0');
+	}
+
+	m_Client->SendPluginMessage("REGISTER", ChannelsString);
+	SendServerHello();
+}
+
+
+
+
+
+void cForgeHandshake::SendServerHello()
+{
+	AString Message;
+	// Discriminator | Byte | Always 0 for ServerHello
+	Message.push_back(Discriminator_ServerHello);
+	// FML protocol Version | Byte | Determined from NetworkRegistery. Currently 2.
+	Message.push_back('\2');
+	// Dimension TODO
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+
+	m_Client->SendPluginMessage("FML|HS", Message);
+}
+
+
+
+
+
+AStringMap cForgeHandshake::ParseModList(const char * a_Data, size_t a_Size)
+{
+	AStringMap Mods;
+
+	if (a_Size < 4)
+	{
+		LOG("ParseModList invalid packet, missing length (size=%zu)", a_Size);
+		return Mods;
+	}
+
+	cByteBuffer Buf(a_Size);
+	Buf.Write(a_Data, a_Size);
+
+	UInt32 NumMods;
+	if (!Buf.ReadVarInt32(NumMods))
+	{
+		LOG("ParseModList failed to read mod count");
+		return Mods;
+	}
+
+	for (size_t i = 0; i < NumMods; ++i)
+	{
+		AString Name, Version;
+		if (!Buf.ReadVarUTF8String(Name))
+		{
+			LOG("ParseModList failed to read mod name at i=%zu", i);
+			break;
+		}
+
+		if (!Buf.ReadVarUTF8String(Version))
+		{
+			LOG("ParseModList failed to read mod version at i=%zu", i);
+			break;
+		}
+
+		Mods.insert({Name, Version});
+
+		LOGD("ParseModList name=%s, version=%s", Name.c_str(), Version.c_str());
+	}
+
+	return Mods;
```

Comment:
I don't think the extreme logging is needed, it would only spam the server console when broken clients (or new Forge versions with different protocol) connect. If you want to keep it, make it `LOGD` (in Debug builds only).

Also, the code is broken into too little pieces needlessly by the empty lines. The entire `for`-loop should be a single block. Basically you should be able to add a comment after each empty line that would describe (in high level) what the next block is doing. Description such as `Read all mod names and versions` is okay, description such as `Push the parsed name and version into the map` is not okay, because the code makes it obvious.